### ISSUE TITLE
Swarco CI fail to run

### ISSUE THREAD

**Issue by emiltin** [rsmp-nordic/rsmp_validator#45]
Swarco tests fail, maybe because it's still trying to use the PR branch, or the rsmp gem version has changed? I've merged  back to master.

/usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/source/git.rb:204:in `rescue in load_spec_files': https://github.com/rsmp-nordic/rsmp.git (at e0d0e7a@e0d0e7a) is not yet checked out. Run `bundle install` first. (Bundler::GitError)
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/source/git.rb:200:in `load_spec_files'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/source/path.rb:105:in `local_specs'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/source/git.rb:168:in `specs'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/lazy_specification.rb:75:in `__materialize__'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/spec_set.rb:83:in `block in materialize'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/spec_set.rb:80:in `map!'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/spec_set.rb:80:in `materialize'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/definition.rb:170:in `specs'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/definition.rb:237:in `specs_for'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/definition.rb:226:in `requested_specs'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/runtime.rb:101:in `block in definition_method'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/runtime.rb:20:in `setup'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler.rb:149:in `setup'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/setup.rb:20:in `block in <top (required)>'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/ui/shell.rb:136:in `with_level'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/ui/shell.rb:88:in `silence'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/setup.rb:20:in `<top (required)>'
	from /usr/lib/ruby/2.7.0/rubygems/core_ext/kernel_require.rb:92:in `require'
	from /usr/lib/ruby/2.7.0/rubygems/core_ext/kernel_require.rb:92:in `require'
/usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/source/path.rb:203:in `load_spec_files': The path `/home/i0davla/.bundle/ruby/2.7.0/rsmp-e0d0e7acca91` does not exist. (Bundler::PathError)
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/source/git.rb:201:in `load_spec_files'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/source/path.rb:105:in `local_specs'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/source/git.rb:168:in `specs'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/lazy_specification.rb:75:in `__materialize__'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/spec_set.rb:83:in `block in materialize'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/spec_set.rb:80:in `map!'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/spec_set.rb:80:in `materialize'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/definition.rb:170:in `specs'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/definition.rb:237:in `specs_for'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/definition.rb:226:in `requested_specs'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/runtime.rb:101:in `block in definition_method'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/runtime.rb:20:in `setup'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler.rb:149:in `setup'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/setup.rb:20:in `block in <top (required)>'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/ui/shell.rb:136:in `with_level'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/ui/shell.rb:88:in `silence'
	from /usr/share/rubygems-integration/all/gems/bundler-2.1.4/lib/bundler/setup.rb:20:in `<top (required)>'
	from /usr/lib/ruby/2.7.0/rubygems/core_ext/kernel_require.rb:92:in `require'
	from /usr/lib/ruby/2.7.0/rubygems/core_ext/kernel_require.rb:92:in `require'

**Comment by emiltin** [comment]
It still seems the test are not run. I don't see anty recent log files in https://otterdahl.org/rsmp_validator_log/.

Is it checking out master branch, or still trying to use a (now deleted) feature branch? It might need to run bundle before each run, as stated in the above error message.

**Comment by otterdahl** [comment]
> It still seems the test are not run. I don't see anty recent log files in https://otterdahl.org/rsmp_validator_log/.
> 
> Is it checking out master branch, or still trying to use a (now deleted) feature branch? It might need to run bundle before each run, as stated in the above error message.

Sorry, the simulator was not running. I've got some problem running the
simulator continuously during an extended period of time.

I'll try to restart it automatically once every 24h and running the test
manually for now.

**Comment by emiltin** [comment]
Does the simulator run under Linux?

**Comment by otterdahl** [comment]
> Does the simulator run under Linux?

The simulator contains parts which are written for MS Windows as well as Linux.

The Windows part runs using Wine and the Linux parts runs using chroot +
qemu-user since they are compiled for a different hardware platform (arm).

The Linux parts adds RSMP support which the original simulator lacks.

All executables are coordinated using a custom shell script and run in a
virtual machine under Linux (amd64)

**Comment by emiltin** [comment]
OK I see.
Did you run the tests? I still don't see anything new on https://otterdahl.org/rsmp_validator_log/

**Comment by otterdahl** [comment]
> Did you run the tests? I still don't see anything new on https://otterdahl.org/rsmp_validator_log/

Fixed. But it's complaining:

```
An error occurred while loading support/spec_helper.
Failure/Error: require 'active_support/time'

LoadError:
  cannot load such file -- active_support/time
# ./spec/support/spec_helper.rb:1:in `require'
# ./spec/support/spec_helper.rb:1:in `<top (required)>'
No examples found.


Finished in 0.00003 seconds (files took 0.08111 seconds to load)
0 examples, 0 failures, 1 error occurred outside of examples
```

**Comment by otterdahl** [comment]
Last time this error occurred it was fixed using the following patch:

https://github.com/rsmp-nordic/rsmp_validator/commit/2d0e02c1517f951246fdd5df17c5bcdc9c72ef49

**Comment by emiltin** [comment]
yeah, fixed with latest commit

**Comment by emiltin** [comment]
Does it work not? I don't see any new log files. (But for some reason, it seems there are now more old log files than yesterday?)

**Comment by otterdahl** [comment]
Found the problem. It should be fixed now. Running the tests manually for now

**Comment by emiltin** [comment]
Ok. How does it usually take before log files show up?

**Comment by otterdahl** [comment]
About 10 minutes. The page https://otterdahl.org/rsmp_validator_log/ seems to report time stamps one hour off, but the filename is correct. The latest log is https://otterdahl.org/rsmp_validator_log/rspec-2021-01-14_09-08-03.log

**Comment by emiltin** [comment]
Alright. There's a lot of failures. 
Some of this seems to be because e.g switch_plan() uses verify_status, but that request a single statusresponse, it doesn't subscribe. If the TLC doesn't respond within the timeout (in this case 20s) the test fails. I think these tests should subscribes instead?

**Comment by emiltin** [comment]
Which commands are expected to cause an immediate effect that can be verified with a status request, and which can take time to have an affect, and must therefore be verified using a status subscription?

Traffic Light Controller | M0001 | Sets functional position
Traffic Light Controller | M0002 | Sets current time plan
Traffic Light Controller | M0003 | Sets traffic situation the controller uses.
Traffic Light Controller | M0004 | Restarts Traffic Light Controller
Traffic Light Controller | M0005 | Activate emergency route
Traffic Light Controller | M0006 | Activate input
Traffic Light Controller | M0007 | Activate fixed time control
Detector logic | M0008 | Sets manual activation of detector logic
Signal group | M0010 | Start of signal group. Orders a signal group to green.
Signal group | M0011 | Stop of signal group. Orders a signal group to red.
Traffic Light Controller | M0012 | Request start or stop of a series of signal groups.
Traffic Light Controller | M0013 | Activate a series of inputs
Traffic Light Controller | M0014 | Set dynamic bands
Traffic Light Controller | M0015 | Set Offset time
Traffic Light Controller | M0016 | Set week time table
Traffic Light Controller | M0017 | Set time tables
Traffic Light Controller | M0018 | Set Cycle time
Traffic Light Controller | M0019 | Force input
Traffic Light Controller | M0020 | Force output
Traffic Light Controller | M0021 | Set trigger level sensitivity for loop detector
Traffic Light Controller | M0103 | Set security code
Traffic Light Controller | M0104 | Set clock

**Comment by emiltin** [comment]
The test for M0104 set date fails because the months is send as "09" but returned as "9". Is there a reasons for sending it as "09" rather than "9"? It would be good to capture the accepted behaviour in our tests.

**Comment by emiltin** [comment]
Can you try running the tests again?

**Comment by emiltin** [comment]
Please rerun tests

**Comment by emiltin** [comment]
In the last run, I noticed the simulator tried to connect twice:
```

Running test ./spec/site/alarm_spec.rb:9 - RSMP site alarm A0302 detector error (logic error)
2021-01-14 09:51:07 UTC                           Connecting
2021-01-14 09:51:07 UTC                           Starting supervisor on port 12111
2021-01-14 09:51:07 UTC                           Waiting for site to connect
2021-01-14 09:51:08 UTC                           Site connected from 127.0.0.1:40076
2021-01-14 09:51:08 UTC  KK+AG9998     -->  2089  Received Version message for site KK+AG9998 using RSMP 3.1.2 {"mType":"rSMsg","type":"Version","mId":"20895de1-770d-4373-afe9-598a4428b473","siteId":[{"sId":"KK+AG9998"}],"RSMP":[{"vers":"3.1.2"}],"SXL":"1.0.7"}
2021-01-14 09:51:08 UTC  KK+AG9998     <--        Sent MessageAck for Version 2089 {"mType":"rSMsg","type":"MessageAck","oMId":"20895de1-770d-4373-afe9-598a4428b473"}
2021-01-14 09:51:08 UTC  KK+AG9998     <--  0323  Sent Version {"mType":"rSMsg","type":"Version","RSMP":[{"vers":"3.1.1"},{"vers":"3.1.2"},{"vers":"3.1.3"},{"vers":"3.1.4"}],"siteId":[{"sId":"KK+AG9998"}],"SXL":"1.0.7","mId":"03235e85-af86-4168-9edb-5edc1279066a"}
2021-01-14 09:51:08 UTC  KK+AG9998     -->  72ea  Received StatusUpdate {"mType":"rSMsg","type":"StatusUpdate","mId":"72ea3338-0b36-4bfe-94d3-6a8e688065dc","ntsOId":"KK+AG9998=001TC000","xNId":"","cId":"KK+AG9998=001TC000","sTs":"2021-01-14T08:51:20.615Z","sS":[{"sCI":"S0001","n":"signalgroupstatus","s":"bbbbbbbbbbbbbbFGG3OFaaBbBba","q":"recent"}]}
2021-01-14 09:51:08 UTC  KK+AG9998     <--        Sent MessageAck for StatusUpdate 72ea {"mType":"rSMsg","type":"MessageAck","oMId":"72ea3338-0b36-4bfe-94d3-6a8e688065dc"}
2021-01-14 09:51:08 UTC  KK+AG9998     -->  4630  Received Watchdog {"mType":"rSMsg","type":"Watchdog","mId":"46300dba-08a3-41a4-ac97-ea770f54419d","wTs":"2021-01-14T09:51:08.832Z"}
2021-01-14 09:51:08 UTC  KK+AG9998     <--        Sent MessageAck for Watchdog 4630 {"mType":"rSMsg","type":"MessageAck","oMId":"46300dba-08a3-41a4-ac97-ea770f54419d"}
2021-01-14 09:51:08 UTC  KK+AG9998     <--  e4c0  Sent Watchdog {"mType":"rSMsg","type":"Watchdog","wTs":"2021-01-14T09:51:08.855Z","mId":"e4c0ccbf-3a38-4044-834e-a4e5a61a01ba"}
2021-01-14 09:51:08 UTC  KK+AG9998     -->  0dac  Received AggregatedStatus status for component KK+AG9998=001TC000 [low_priority_alarm, normal] {"mType":"rSMsg","type":"AggregatedStatus","mId":"0daca855-7ee7-4d97-bd1c-e7af2cd3e165","ntsOId":"KK+AG9998=001TC000","xNId":"","cId":"KK+AG9998=001TC000","aSTS":"2021-01-14T09:51:08.833Z","fP":"","fS":"","se":[false,false,false,false,true,true,false,false]}
2021-01-14 09:51:08 UTC  KK+AG9998     <--        Sent MessageAck for AggregatedStatus 0dac {"mType":"rSMsg","type":"MessageAck","oMId":"0daca855-7ee7-4d97-bd1c-e7af2cd3e165"}
2021-01-14 09:51:08 UTC  KK+AG9998     -->        Received MessageAck for Version 0323 {"mType":"rSMsg","type":"MessageAck","oMId":"03235e85-af86-4168-9edb-5edc1279066a"}
2021-01-14 09:51:08 UTC  KK+AG9998                Connection to site KK+AG9998 established
2021-01-14 09:51:08 UTC  KK+AG9998     -->        Received MessageAck for Watchdog e4c0 {"mType":"rSMsg","type":"MessageAck","oMId":"e4c0ccbf-3a38-4044-834e-a4e5a61a01ba"}
2021-01-14 09:51:32 UTC  KK+AG9998                Waiting for alarm
2021-01-14 09:51:32 UTC                           Site connected from 127.0.0.1:40078
2021-01-14 09:51:32 UTC                -->  6f63  Rejected Version, Site KK+AG9998 already connected {"mType":"rSMsg","type":"Version","mId":"6f63ced0-1ac2-4c43-b27d-fd8b77483aca","siteId":[{"sId":"KK+AG9998"}],"RSMP":[{"vers":"3.1.2"}],"SXL":"1.0.7"}
```

Notice the last line: Rejected Version, Site KK+AG9998 already connected. It first connects from port 40076, which succeeds but then tries again on port 40078. Could this be a kink in the test setup, or a bug in the simulator?

**Comment by emiltin** [comment]
Could two instances of the simulator be running?

**Comment by otterdahl** [comment]
> In the last run, I noticed the simulator tried to connect twice:
...
> Notice the last line: Rejected Version, Site KK+AG9998 already connected. It first connects from port 40076, which succeeds but then tries again on port 40078. Could this be a kink in the test setup, or a bug in the simulator?

I believe it is related to the problems we saw in https://github.com/rsmp-nordic/rsmp_validator/issues/38. Upon establishing connection, the simulator sends old buffered messages, including alarms. `rsmp_validator` responds with NotAck due to this issue: https://github.com/rsmp-nordic/rsmp_implementations/issues/4. When the Swarco simulator receieves NotAck it closes the connection after a while and tries to reconnect - I believe this is causing the multiple connection attempts we see in the log.

> Could two instances of the simulator be running?

There is a check of the PID in the script that should prevent this from happening

**Comment by otterdahl** [comment]
> Which commands are expected to cause an immediate effect that can be verified with a status request, and which can take time to have an affect, and must therefore be verified using a status subscription?

It depends on the TLC. For instance, we know that that M0002 is instantaneous in some TLC's and takes a while in others. I suggest being on the safe side and use status subscription in most cases.

**Comment by otterdahl** [comment]
> The test for M0104 set date fails because the months is send as "09" but returned as "9". Is there a reasons for sending it as "09" rather than "9"? It would be good to capture the accepted behaviour in our tests.

Well, in https://rsmp-nordic.github.io/rsmp_sxl_traffic_lights/sxl_traffic_light_controller.html#m0104 is stated that **Month** should be written as "MM" (and not "M"). But i guess "M" is quite OK too. What do you think?

**Comment by otterdahl** [comment]
> Some of this seems to be because e.g switch_plan() uses verify_status, but that request a single statusresponse, it doesn't subscribe. If the TLC doesn't respond within the timeout (in this case 20s) the test fails. I think these tests should subscribes instead?

Yes, 20s is far too short for switch_plan(). It may take a whole cycle until
the plan is switched in the TLC. A cycle can be everything from 20s up to 80s
or longer. Perhaps set the timeout to 200s?

The simulator should switch time plan very quickly, so it shouldn't be a
problem in this case. We have to look into what is causing the issue.

**Comment by emiltin** [comment]
Well, the spec shows 09 must be used, and the status uses the same format. I think we should require two digits in the schema, and flag the Swarco behaviour as a bug.

**Comment by emiltin** [comment]
Can you rerun the tests? I've added the port to the logging, so it's easier to see what's going on with the multiple connections.

**Comment by emiltin** [comment]
It's now command_timeout from the config, which is currently set to 180 seconds.

**Comment by emiltin** [comment]
Is the simulator restarted between test runs?

**Comment by otterdahl** [comment]
> Can you rerun the tests? I've added the port to the logging, so it's easier to see what's going on with the multiple connections.

Sure. Hold on. The 'git pull' failed in the script due to changes to
Gemfile.lock. The test continues anyway, but without the updated repo.

Will fix this and rerun the test

**Comment by otterdahl** [comment]
> Is the simulator restarted between test runs?

No. It runs all the time.

To start it requires `sudo` for chroot + qemu-user, so it need to be started
manually. Therefore it keeps the old buffered messages.

**Comment by emiltin** [comment]
I also see subscriptions continuing. What effect will a restart via RSMP have?

**Comment by emiltin** [comment]
Do you run bundle on each run? That will ensure you have the right version of the gem

**Comment by otterdahl** [comment]
> I also see subscriptions continuing. What effect will a restart via RSMP have?

Do you mean restart using M0004? It is not expected to have any effect (since
we're using a simulator).

**Comment by otterdahl** [comment]
> Do you run bundle on each run? That will ensure you have the right version of the gem

Not at the moment. I believe it needs `sudo` sometimes, so I run it manually
from time to time. Perhaps this can be fixed.

**Comment by emiltin** [comment]
Some tests abort due to missing scripts:

```
  1) RSMP site alarm A0302 detector error (logic error)
     Failure/Error: raise "Aborting test because script config is missing" unless SCRIPT_PATHS

     RuntimeError:
       Aborting test because script config is missing
```

The shell command to run (with full path must be in the config under 'scripts'.

**Comment by otterdahl** [comment]
> The shell command to run (with full path must be in the config under 'scripts'.

Fixed

**Comment by emiltin** [comment]
I fixed a few other things.

**Comment by emiltin** [comment]
It does looks like two instances of simulator is trying to connect, or the simulator tries to connect on from two different ports. The seconds connect is reject because it's using the same site id as the one already connected.

**Comment by otterdahl** [comment]
> It does looks like two instances of simulator is trying to connect, or the simulator tries to connect on from two different ports. The seconds connect is reject because it's using the same site id as the one already connected.

That's strange. Only once instance of the simulator is running as far as I can
see

**Comment by emiltin** [comment]
Maybe we should try restarting the simulator? It seems like it might be misbehaving.

**Comment by otterdahl** [comment]
> Maybe we should try restarting the simulator? It seems like it might be misbehaving.

Yes. I was working on the simulator to fix the time sync. It should be up now

**Comment by emiltin** [comment]
Can you run the test, please?

**Comment by otterdahl** [comment]
> Can you run the test, please?

Sure. Running now

**Comment by emiltin** [comment]
It's still connecting on twice on different ports:
```
2021-01-15 08:48:22 UTC  35012    KK+AG9998     <--  1aea  Sent Watchdog {"mType":"rSMsg","type":"Watchdog","wTs":"2021-01-15T08:48:22.770Z","mId":"1aea6192-2b43-4f6b-b2e9-a7af0eb6b19d"}
2021-01-15 08:48:22 UTC  35012    KK+AG9998     -->        Received MessageAck for Watchdog 1aea {"mType":"rSMsg","type":"MessageAck","oMId":"1aea6192-2b43-4f6b-b2e9-a7af0eb6b19d"}
2021-01-15 08:48:22 UTC  35012    KK+AG9998     -->  3d24  Received Watchdog {"mType":"rSMsg","type":"Watchdog","mId":"3d245683-7dbb-4670-b37e-7edb16ae3268","wTs":"2021-01-15T08:48:22.777Z"}
2021-01-15 08:48:22 UTC  35012    KK+AG9998     <--        Sent MessageAck for Watchdog 3d24 {"mType":"rSMsg","type":"MessageAck","oMId":"3d245683-7dbb-4670-b37e-7edb16ae3268"}

Running test ./spec/site/alarm_spec.rb:40 - RSMP site alarm Acknowledge alarm
2021-01-15 08:49:11 UTC  35012    KK+AG9998                Waiting for alarm
2021-01-15 08:49:11 UTC  35034                             Site connected from 127.0.0.1:35034
2021-01-15 08:49:11 UTC  35034                  -->  3667  Rejected Version, Site KK+AG9998 already connected {"mType":"rSMsg","type":"Version","mId":"36672d0b-410b-4c25-bf2b-32c96f3bcfd3","siteId":[{"sId":"KK+AG9998"}],"RSMP":[{"vers":"3.1.2"}],"SXL":"1.0.7"}
2021-01-15 08:49:11 UTC  35034                  <--        Sent MessageNotAck for Version 3667 {"mType":"rSMsg","type":"MessageNotAck","rea":"Site KK+AG9998 already connected","oMId":"36672d0b-410b-4c25-bf2b-32c96f3bcfd3"}
2021-01-15 08:49:11 UTC  35034                             Closing connection to site
```

Weird.

**Comment by otterdahl** [comment]
> It's still connecting twice on different ports:

Yes. As I mentioned earlier, I think is due to the simulator disconnecting and reconnecting. In fact, I'm quite sure of it. It is disconnecting due to receiving a NotAck from rsmp_validator since the simulator does not comply to the rsmp_schema (see rsmp-nordic/rsmp_implementations#4).

The simulator has it's own log file. When the last test was run the following was logged:
```
(2021-01-15 08:48:57) : [INFO]   No reply received from server. Closing connection
(2021-01-15 08:48:57) : [INFO]   Connection to server terminated. Shutting down
```
and then
```
(2021-01-15 08:48:57) : [INFO]   Connection to server established. Starting up
```
and then finally
```
(2021-01-15 08:49:11) : [INFO]   NACK received. [ID: 36672d0b-410b-4c25-bf2b-32c96f3bcfd3]
(2021-01-15 08:49:11) : [INFO]   RECEIVING : {"mType":"rSMsg","type":"MessageNotAck","rea":"Site KK+AG9998 already connected","oMId":"36672d0b-410b-4c25-bf2b-32c96f3bcfd3"}^L

(2021-01-15 08:49:11) : [INFO]   Connection to server terminated. Shutting down
```
The reason to why didn't see this earlier is due to several reasons:
* The simulator is running continuously and have buffered messages, incl. alarms. These are sent upon establishment and rejected due to rsmp-nordic/rsmp_implementations#4). 
* It doesn't close the connection and reconnects immediately. There is a timeout of about 30 seconds. If the simulator is reset and buffered alarms cleared, the first test case may pass without error (since the simulators haven't timed out yet)
* We haven't been able to test alarms earlier

**Comment by emiltin** [comment]
Not sure I understand why this would cause the simulator to open tow connections to the supervisor at the same time.
Could the simulator logs be published on the server as well? I guess you would have to rotate the log file on each test run.

**Comment by emiltin** [comment]
The Ruby supervisor rejects the second connect and says it will close the connection. But a number messages are received before it's actually closed. This is also a bit odd and I will look into that. It would be better to immediately close the socket and ignore any additional incoming message, I think.

**Comment by otterdahl** [comment]
> Could the simulator logs be published on the server as well? I guess you would have to rotate the log file on each test run. 

I'm on it. They will be published as rsmp_log_simulator_(date).log

**Comment by otterdahl** [comment]
First log published: https://otterdahl.org/rsmp_validator_log/rsmp_log_simulator-2021-01-15_15-27-21.log

**Comment by emiltin** [comment]
In the latest run, it does not connect twice, even though invalid Alarms are still NotAcked by the rsmp validator. So I'm not sure NotAcks are the cuase of double connections.

**Comment by emiltin** [comment]
Did you find out more about why it's rejecting component id KK+AG9998=001SG1? Eg:

```
2021-01-15 14:26:31 UTC  54570    KK+AG9998     <--  618e  Sent CommandRequest {"mType":"rSMsg","type":"CommandRequest","ntsOId":"","xNId":"","cId":"KK+AG9998=001SG1","arg":[{"cCI":"M0010","cO":"setStart","n":"securityCode","v":""},{"cCI":"M0010","cO":"setStart","n":"status","v":"True"}],"mId":"618ec89c-8ef8-463e-8925-1750bfa35aba"}
2021-01-15 14:26:31 UTC  54570    KK+AG9998     -->        Received MessageNotAck for CommandRequest 618e: Component ID not recognized. {"mType":"rSMsg","type":"MessageNotAck","oMId":"618ec89c-8ef8-463e-8925-1750bfa35aba","rea":"Component ID not recognized."}
```

The simulator log is a bit akward because it shows the sending of the NotAck before the reception of the request:

```
(2021-01-15 14:26:31) : [INFO]	 Received Command Request
(2021-01-15 14:26:31) : [INFO]	 Controller Code : KK+AG9998=001TC000.
(2021-01-15 14:26:31) : [INFO]	 SENDING : {"mType":"rSMsg","type":"MessageNotAck","oMId":"618ec89c-8ef8-463e-8925-1750bfa35aba","rea":"Component ID not recognized."}

(2021-01-15 14:26:31) : [INFO]	 Command Request is invalid
(2021-01-15 14:26:31) : [INFO]	 RECEIVING : {"mType":"rSMsg","type":"CommandRequest","ntsOId":"","xNId":"","cId":"KK+AG9998=001SG1","arg":[{"cCI":"M0010","cO":"setStart","n":"securityCode","v":""},{"cCI":"M0010","cO":"setStart","n":"status","v":"True"}],"mId":"618ec89c-8ef8-463e-8925-1750bfa35aba"}
```

**Comment by otterdahl** [comment]
> Did you find out more about why it's rejecting component id KK+AG9998=001SG1? Eg:

This version of the simulator is a few years old and doesn't support M0010.
It reports "Component ID not recognized." but this error message seems
misleading. I can't find anything wrong with the component id.

> The simulator log is a bit akward because it shows the sending of the NotAck before the reception of the request:

Agree. This seems a bit mixed up

**Comment by emiltin** [comment]
Considering the age of the simulator, which of these failed tests is expected to fail, and which indicate a bug in either of validator or the simulator?

```
rspec ./spec/site/alarm_spec.rb:9 # RSMP site alarm A0302 detector error (logic error)
rspec ./spec/site/command_spec.rb:16 # RSMP site commands M0001 set dark mode
rspec ./spec/site/command_spec.rb:38 # RSMP site commands M0004 restart
rspec ./spec/site/command_spec.rb:86 # RSMP site commands M0010 start signal group
rspec ./spec/site/command_spec.rb:93 # RSMP site commands M0011 stop signal group
rspec ./spec/site/command_spec.rb:100 # RSMP site commands M0012 request start/stop of a series of signal groups
rspec ./spec/site/command_spec.rb:115 # RSMP site commands M0014 set command table
rspec ./spec/site/command_spec.rb:124 # RSMP site commands M0015 set offset
rspec ./spec/site/command_spec.rb:133 # RSMP site commands M0016 set week table
rspec ./spec/site/command_spec.rb:141 # RSMP site commands M0017 set time table
rspec ./spec/site/command_spec.rb:149 # RSMP site commands M0018 set cycle time
rspec ./spec/site/command_spec.rb:158 # RSMP site commands M0019 force input
rspec ./spec/site/command_spec.rb:168 # RSMP site commands M0020 force output
rspec ./spec/site/command_spec.rb:178 # RSMP site commands M0021 set trigger sensitivity
rspec ./spec/site/command_spec.rb:203 # RSMP site commands Send the wrong security code
rspec ./spec/site/connect_spec.rb:32 # RSMP site connection Version 3.1.2 exchanges correct connection sequence
rspec ./spec/site/connect_spec.rb:47 # RSMP site connection Version 3.1.3 exchanges correct connection sequence
rspec ./spec/site/connect_spec.rb:64 # RSMP site connection Version 3.1.4 exchanges correct connection sequence
rspec ./spec/site/status_spec.rb:109 # RSMP site status S0022 list of time plans
rspec ./spec/site/status_spec.rb:114 # RSMP site status S0023 command table
rspec ./spec/site/status_spec.rb:119 # RSMP site status S0024 offset time
rspec ./spec/site/status_spec.rb:124 # RSMP site status S0025 time-of-green/time-of-red
rspec ./spec/site/status_spec.rb:138 # RSMP site status S0026 week time table
rspec ./spec/site/status_spec.rb:143 # RSMP site status S0027 time tables
rspec ./spec/site/status_spec.rb:148 # RSMP site status S0028 cycle time
rspec ./spec/site/status_spec.rb:153 # RSMP site status S0029 forced input status
rspec ./spec/site/status_spec.rb:158 # RSMP site status S0030 forced output status
rspec ./spec/site/status_spec.rb:163 # RSMP site status S0031 trigger level sensitivity for loop detector
rspec ./spec/site/status_spec.rb:168 # RSMP site status S0091 operator logged in/out OP-panel
rspec ./spec/site/status_spec.rb:173 # RSMP site status S0092 operator logged in/out web-interface
rspec ./spec/site/status_spec.rb:195 # RSMP site status S0097 version of traffic program
rspec ./spec/site/status_spec.rb:206 # RSMP site status S0202 traffic counting: vehicle speed
rspec ./spec/site/status_spec.rb:218 # RSMP site status S0204 traffic counting: classification
rspec ./spec/site/status_spec.rb:235 # RSMP site status S0205 traffic counting: number of vehicles
rspec ./spec/site/status_spec.rb:240 # RSMP site status S0206 traffic counting: vehicle speed
rspec ./spec/site/status_spec.rb:245 # RSMP site status S0207 traffic counting: occupancy
rspec ./spec/site/status_spec.rb:250 # RSMP site status S0208 traffic counting: classification
```

My goal is en ensure that the validator correctly identifies real issues.

**Comment by otterdahl** [comment]
> Considering the age of the simulator, which of these failed tests is expected to fail, and which indicate a bug in either of validator or the simulator?

Officially the simulator only support SXL 1.0.7

The following are part of SXL 1.0.7 and should work:

> ```
> rspec ./spec/site/alarm_spec.rb:9 # RSMP site alarm A0302 detector error (logic error)
> rspec ./spec/site/command_spec.rb:16 # RSMP site commands M0001 set dark mode
> rspec ./spec/site/command_spec.rb:38 # RSMP site commands M0004 restart
> rspec ./spec/site/command_spec.rb:203 # RSMP site commands Send the wrong security code
> rspec ./spec/site/connect_spec.rb:32 # RSMP site connection Version 3.1.2 exchanges correct connection sequence
> rspec ./spec/site/status_spec.rb:168 # RSMP site status S0091 operator logged in/out OP-panel
> rspec ./spec/site/status_spec.rb:173 # RSMP site status S0092 operator logged in/out web-interface
> rspec ./spec/site/status_spec.rb:206 # RSMP site status S0202 traffic counting: vehicle speed
> rspec ./spec/site/status_spec.rb:218 # RSMP site status S0204 traffic counting: classification
> ```

But please note that Swarco has declared that they don't support S0091, S0092,
S0202 and S0204. Also note that M0004 only works in a real controller.

The following were added after SXL 1.0.7 and is not expected to work:
> ```
> rspec ./spec/site/command_spec.rb:86 # RSMP site commands M0010 start signal group
> rspec ./spec/site/command_spec.rb:93 # RSMP site commands M0011 stop signal group
> rspec ./spec/site/command_spec.rb:100 # RSMP site commands M0012 request start/stop of a series of signal groups
> rspec ./spec/site/command_spec.rb:115 # RSMP site commands M0014 set command table
> rspec ./spec/site/command_spec.rb:124 # RSMP site commands M0015 set offset
> rspec ./spec/site/command_spec.rb:133 # RSMP site commands M0016 set week table
> rspec ./spec/site/command_spec.rb:141 # RSMP site commands M0017 set time table
> rspec ./spec/site/command_spec.rb:149 # RSMP site commands M0018 set cycle time
> rspec ./spec/site/command_spec.rb:158 # RSMP site commands M0019 force input
> rspec ./spec/site/command_spec.rb:168 # RSMP site commands M0020 force output
> rspec ./spec/site/command_spec.rb:178 # RSMP site commands M0021 set trigger sensitivity
> rspec ./spec/site/connect_spec.rb:47 # RSMP site connection Version 3.1.3 exchanges correct connection sequence
> rspec ./spec/site/connect_spec.rb:64 # RSMP site connection Version 3.1.4 exchanges correct connection sequence
> rspec ./spec/site/status_spec.rb:109 # RSMP site status S0022 list of time plans
> rspec ./spec/site/status_spec.rb:114 # RSMP site status S0023 command table
> rspec ./spec/site/status_spec.rb:119 # RSMP site status S0024 offset time
> rspec ./spec/site/status_spec.rb:124 # RSMP site status S0025 time-of-green/time-of-red
> rspec ./spec/site/status_spec.rb:138 # RSMP site status S0026 week time table
> rspec ./spec/site/status_spec.rb:143 # RSMP site status S0027 time tables
> rspec ./spec/site/status_spec.rb:148 # RSMP site status S0028 cycle time
> rspec ./spec/site/status_spec.rb:153 # RSMP site status S0029 forced input status
> rspec ./spec/site/status_spec.rb:158 # RSMP site status S0030 forced output status
> rspec ./spec/site/status_spec.rb:163 # RSMP site status S0031 trigger level sensitivity for loop detector
> rspec ./spec/site/status_spec.rb:195 # RSMP site status S0097 version of traffic program
> rspec ./spec/site/status_spec.rb:235 # RSMP site status S0205 traffic counting: number of vehicles
> rspec ./spec/site/status_spec.rb:240 # RSMP site status S0206 traffic counting: vehicle speed
> rspec ./spec/site/status_spec.rb:245 # RSMP site status S0207 traffic counting: occupancy
> rspec ./spec/site/status_spec.rb:250 # RSMP site status S0208 traffic counting: classification
> ```

**Comment by otterdahl** [comment]
> In the latest run, it does not connect twice, even though invalid Alarms are still NotAcked by the rsmp validator. So I'm not sure NotAcks are the cuase of double connections.

True.

Between the runs, the simulator were restarted and any buffered alarms and statuses were cleared. This might have changed the outcome.

Anyway, when experiencing duplicate connections the simulator complains
about not receiving a reply for some reason:

```
(2021-01-15 08:48:57) : [INFO]   No reply received from server. Closing connection
(2021-01-15 08:48:57) : [INFO]   Connection to server terminated. Shutting down
```

**Comment by otterdahl** [comment]
Should I try to clear the buffer of the simulator between runs?

Perhaps we should add a specific test case to test the buffer in the future

**Comment by emiltin** [comment]
Thanks. 
Yes if possible, clear the buffer between runs. The more we can ensure a consistent start state, the better. When testing physical equipment, this will often not be possible, but for now it will make it easier for us to identify and fix any errors in the validator.

Around the mesage where the simulator says No reply from server :
(2021-01-15 08:48:57) : [INFO]   No reply received from server. Closing connection

I don't see anything in the validator logs. But earlier it rejected a double connection:
```
2021-01-18 09:12:15 UTC  60470                  -->  20a1  Rejected Version, Site KK+AG9998 already connected {"mType":"rSMsg","type":"Version","mId":"20a1d81d-1d6f-4cf6-9197-98d018528579","siteId":[{"sId":"KK+AG9998"}],"RSMP":[{"vers":"3.1.2"}],"SXL":"1.0.7"}
```
After that the log repeatedly says it's closing the connection. Maybe the validator stops reading messages, but doesn't fully terminate the connection. That's why the simulator at some point ways that no response was received.

**Comment by otterdahl** [comment]
> Yes if possible, clear the buffer between runs. The more we can ensure a consistent start state, the better. When testing physical equipment, this will often not be possible, but for now it will make it easier for us to identify and fix any errors in the validator.

Done. Hopefully it works

**Comment by emiltin** [comment]
Thanks.
It seems the swarco simulator might not rerun if commits are added while it's already running?

**Comment by emiltin** [comment]
Closing this issue as the origin issue is solved. Let's open new issue as needed

**Comment by otterdahl** [comment]
> It seems the swarco simulator might not rerun if commits are added while it's already running?

True. It won't rerun if it is already running. I'll try to fix this
